Improve PCA performance with irlba

daniel-wells opened the pull request:
Hi Jesse,

I finally got round to implementing a suggestion to resolve issue #26 

I've added an option partial_pca which when set to TRUE uses irlba to compute first n principal components. The default is FALSE so no breaking changes.

I added the irlba as a suggested package and then check before using it so no hard dependencies are added.

jkrijthe commented:
Thanks! I worked on a local implementation as well, and was considering making it the default behaviour, but did not get around to testing it yet. Anyway, your suggestion looks better! Two requests:
- I would prefer not to have the message generated for large matrices when non-partial pca is used, since this would give warnings for people who don't currently, and they would not be able to turn the messages off.
- can you add 1 or 2 simple unit tests? Perhaps at least checking that solutions are similar when partial_pca=TRUE vs FALSE?

daniel-wells commented:
I changed the message so it would only output when verbose=TRUE and coloured black so it looks less like a warning. If you'd rather it just be removed completely just let me know and I'll remove.

I added a couple of tests, one which tests the solutions are similar (although due to the principal components having arbitrary sign I had to do this via the tsne costs unless you wanted to also [optionally] return the pca / the pca standard deviations). Another test checks that irlba is actually faster than pca.

jkrijthe commented:
Thanks! Final changes :-): the [R extensions documentation](ran.r-project.org/doc/manuals/r-release/R-exts.html)  advises against tests based on timing (section 1.6), so I would suggest removing the timing tests. For the other test, please also set the seed before generating the data (section 1.6).

daniel-wells commented:
Good catch, I forgot GC could mess with the timings. I've implemented the requested changes.
